backend/app/api/api_v1/endpoints/expectation.py

```diff
@@ -172,16 +176,38 @@ def delete_expectation(
     )
 
 
-def zip_expectations_and_validations(expectations: list[Expectation], validations):
-
+def zip_expectations_and_validations(expectations: list[Expectation], validations: list[Validation]):
     expectations_as_dict: dict[str, Expectation] = {}
 
     for expectation in expectations:
         expectations_as_dict[expectation.key] = expectation
 
-    for v in validations:
-        source = v["_source"]
-        source["run_date"] = utils.string_to_military_time(source["run_date"])
-        expectations_as_dict[source["expectation_id"]].validations.append(source)
+    for validation in validations:
+        run_time = utils.string_to_military_time(validation.meta.run_id.run_time)
+
+        for result in validation.results:
+            result_as_dict = result.dict()
+            result_as_dict["run_time"] = run_time
+
+            # when an expectation is deleted, the corresponding validation results are not deleted. Instead, validations
+            # should be expired/ deleted after some portion of time. For this reason, we only want to zip/join
+            # a validation result to an expectation that hasn't been deleted.
+            if expectations_as_dict.get(result.expectation_id):
+                expectations_as_dict[result.expectation_id].validations.append(result_as_dict)
 
     return list(expectations_as_dict.values())
+
+
+def _table_level_expectation_already_exists(expectation: Expectation, repository: ExpectationRepository):
+    # Duplicate Table level/ result_type="expectation", expectations are removed by GE when validations are run.
+    # Because of this, we want to prevent duplicate table level expectations from being added.
+    if expectation.result_type == c.EXPECTATION:
+        if len(repository.query_by_filter(
```

Comment:
Could be interesting in terms of performance to use `count` here.

What we could do is to factorize the query build into a private method of the repository and use it both in `query_by_filter` and  a new method `count_by_filter`.

Reply:
I like this. I implemented the filtering logic in `_build_query_filter` which `delete_by_filter` and `count_by_filter` now use.

I think we will want to create some "query builder" in `base` repository if the number of filters grow any more.